From e1e4a635f8bf7c1e2ccd0d3949b254b782cee449 Mon Sep 17 00:00:00 2001
From: Peter Varga <pvarga@inf.u-szeged.hu>
Date: Wed, 14 May 2025 14:02:56 +0200
Subject: [PATCH 09/13] UPSTREAM: Make possible to use fake GL API with Vulkan

This configuration disables ANGLE but it still renders with Vulkan.

Enable it with:
--webEngineArgs --enable-features=Vulkan --use-vulkan=native
--use-gl=stub

WebGL is not expected to work with this configuration since it requires
real GL API.

Pick-to: 6.10
Change-Id: I7bc292e0c131c2cebe345f70bd3163dd3e6b2972
Reviewed-by: Moss Heim <moss.heim@qt.io>
---
 qtwebengine/src/core/compositor/display_overrides.cpp |  3 ---
 qtwebengine/src/core/ozone/surface_factory_qt.cpp     |  1 +
 qtwebengine/src/core/web_engine_context.cpp           | 11 +++++++++--
 3 files changed, 10 insertions(+), 5 deletions(-)

diff --git a/qtwebengine/src/core/compositor/display_overrides.cpp b/qtwebengine/src/core/compositor/display_overrides.cpp
index 32012f7cf3..c33e79e54e 100644
--- a/qtwebengine/src/core/compositor/display_overrides.cpp
+++ b/qtwebengine/src/core/compositor/display_overrides.cpp
@@ -40,9 +40,6 @@ viz::SkiaOutputSurfaceImplOnGpu::CreateOutputDevice()
 
 #if QT_CONFIG(opengl)
     if (graphicsApi == QSGRendererInterface::OpenGL) {
-        if (gl::GetGLImplementation() != gl::kGLImplementationEGLANGLE)
-            qFatal("OpenGL is only supported over ANGLE.");
-
         return std::make_unique<QtWebEngineCore::NativeSkiaOutputDeviceOpenGL>(
                 context_state_, renderer_settings_.requires_alpha_channel,
                 shared_gpu_deps_->memory_tracker(), dependency_.get(), shared_image_factory_.get(),
diff --git a/qtwebengine/src/core/ozone/surface_factory_qt.cpp b/qtwebengine/src/core/ozone/surface_factory_qt.cpp
index 95af160a72..05f5b43f7e 100644
--- a/qtwebengine/src/core/ozone/surface_factory_qt.cpp
+++ b/qtwebengine/src/core/ozone/surface_factory_qt.cpp
@@ -39,6 +39,7 @@ SurfaceFactoryQt::SurfaceFactoryQt()
     m_impls.push_back({ gl::GLImplementationParts(gl::kGLImplementationEGLANGLE),
                         std::make_unique<ui::GLOzoneANGLEQt>() });
 #endif
+    m_impls.push_back({ gl::GLImplementationParts(gl::kGLImplementationStubGL), nullptr });
     m_impls.push_back({ gl::GLImplementationParts(gl::kGLImplementationDisabled), nullptr });
 }
 
diff --git a/qtwebengine/src/core/web_engine_context.cpp b/qtwebengine/src/core/web_engine_context.cpp
index 074accb6cf..9f11b39951 100644
--- a/qtwebengine/src/core/web_engine_context.cpp
+++ b/qtwebengine/src/core/web_engine_context.cpp
@@ -408,8 +408,15 @@ static std::string getGLType(const base::CommandLine &cmd)
     return gl::kGLImplementationANGLEName;
 }
 
-static bool isGLTypeSupported(const std::string &glType)
+static bool isGLTypeSupported(const std::string &glType, bool usingVulkan = false)
 {
+#if BUILDFLAG(IS_OZONE)
+    if (glType == gl::kGLImplementationStubName)
+        return usingVulkan;
+#else
+    Q_UNUSED(usingVulkan);
+#endif
+
     if (glType == gl::kGLImplementationANGLEName || glType == gl::kGLImplementationDisabledName)
         return true;
 
@@ -1001,7 +1008,7 @@ WebEngineContext::WebEngineContext()
     logContext(parsedCommandLine);
 
     // Early error on unsupported --use-gl settings.
-    if (!isGLTypeSupported(glType))
+    if (!isGLTypeSupported(glType, isFeatureEnabled(features::kVulkan.name, parsedCommandLine)))
         qFatal("--use-gl=%s is not supported with the current configuration.", glType.c_str());
 
     registerMainThreadFactories();
-- 
2.51.0

